### ISSUE TITLE
`productElementNames` throw an `java.lang.IndexOutOfBoundsException: 0` exception

### ISSUE THREAD

**Issue by arturaz** [lampepfl/dotty#17146]
## Compiler version

3.3.0-RC3

## Minimized code

`build.sbt`:
```scala
libraryDependencies += "com.typesafe.play" %% "play-json" % "2.10.0-RC7"
```

Code:
```scala
import play.api.libs.json.*

try { 
  // This throws an `java.lang.IndexOutOfBoundsException: 0` exception.
  JsFalse.productElementNames.toVector
}
catch { case t =>
  t.printStackTrace()
}
```

https://scastie.scala-lang.org/qNqSm2IgTPaMrfSFfY5i8w

## Output

<details>

```scala
java.lang.IndexOutOfBoundsException: 0
	at play.api.libs.json.JsFalse$.productElementName(JsValue.scala:73)
	at scala.Product$$anon$2.next(Product.scala:70)
	at scala.Product$$anon$2.next(Product.scala:66)
	at scala.collection.mutable.Growable.addAll(Growable.scala:62)
	at scala.collection.mutable.Growable.addAll$(Growable.scala:57)
	at scala.collection.immutable.VectorBuilder.addAll(Vector.scala:1604)
	at scala.collection.immutable.Vector$.from(Vector.scala:1388)
	at scala.collection.IterableOnceOps.toVector(IterableOnce.scala:1290)
	at scala.collection.IterableOnceOps.toVector$(IterableOnce.scala:1290)
	at scala.collection.AbstractIterator.toVector(Iterator.scala:1300)
	at Playground$.liftedTree1$1(main.scala:7)
	at Playground$.<clinit>(main.scala:11)
	at Main$.<clinit>(main.scala:15)
	at Main.main(main.scala)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at sbt.Run.invokeMain(Run.scala:143)
	at sbt.Run.execute$1(Run.scala:93)
	at sbt.Run.$anonfun$runWithLoader$5(Run.scala:120)
	at sbt.Run$.executeSuccess(Run.scala:186)
	at sbt.Run.runWithLoader(Run.scala:120)
	at sbt.Run.run(Run.scala:127)
	at com.olegych.scastie.sbtscastie.SbtScastiePlugin$$anon$1.$anonfun$run$1(SbtScastiePlugin.scala:38)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at sbt.util.InterfaceUtil$$anon$1.get(InterfaceUtil.scala:21)
	at sbt.ScastieTrapExit$App.run(ScastieTrapExit.scala:258)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

</details>

## Expectation

Shouldn't crash.

**Comment by SethTisue** [comment]
We need to rule out the possibility that the bug is in the external dependency, rather than in Scala 3 itself. Have you attempted to construct a standalone reproduction?

**Comment by JD557** [comment]
Seems like a library bug, see https://github.com/playframework/play-json/blob/2.10.0-RC7/play-json/shared/src/main/scala/play/api/libs/json/JsValue.scala#L45-L79 and https://scastie.scala-lang.org/ukz89MqpTdip9cAvnLYl9w

However, the default behavior can be quite surprising. I wonder if the default for `case object` should be different.

**Comment by SethTisue** [comment]
But why is the Scala 3 behavior different from the Scala 2 behavior?

**Comment by prolativ** [comment]
Dependency-free minimization:

```scala
abstract class JsBoolean(val value: Boolean) extends Product with Serializable {
  def canEqual(that: Any): Boolean = that.isInstanceOf[JsBoolean]
  val productArity = 1
  def productElement(n: Int): Any =
    if (n == 0) value else throw new IndexOutOfBoundsException(n.toString)
}

case object JsFalse extends JsBoolean(false)
case object JsTrue extends JsBoolean(true)

object Main {
  def main(args: Array[String]): Unit = {
    println(JsFalse.productElementNames.toVector)
  }
}
```

Claude's explanation of the behaviour:

===============================
JsBoolean manually sets productArity = 1 (for backwards compat) but does NOT define productElementName.
JsFalse is a case object extending JsBoolean; Scala 3 synthesizes productElementName for it
(throwing for any n since case objects have 0 fields).
Result: productElementNames uses productArity=1 but productElementName(0) throws.

**Scala 2.13** synthesizes `productElementNames` *directly* for case classes/objects — it emits something equivalent to:

// synthesized by scalac for case object JsFalse
override def productElementNames: Iterator[String] = Iterator.empty
This bypasses productArity entirely. JsFalse has 0 own fields, so the direct synthesized method returns an empty iterator, regardless of the inherited val productArity = 1.

**Scala 3** synthesizes productElementName(n) for the case object (throwing for any n, since it has 0 own fields), but does not synthesize productElementNames — it leaves that to the default Product implementation, which does:

```scala
// Product.scala default — what Scala 3 falls back to
def productElementNames: Iterator[String] = new AbstractIterator {
  val cmax = productArity  // ← picks up the inherited val = 1
  def next() = productElementName(c)  // ← calls synthesized method, throws at n=0
}
```

You can see this in the stack trace — scala.Product$$anon$2 is that anonymous class from the default Product implementation. If Scala 3 had synthesized productElementNames directly (like 2.13 does), it would never reach the productArity-based loop.

In short: Scala 2.13 short-circuits the problem by synthesizing a direct productElementNames. Scala 3's synthesis gap means the loop runs and hits the mismatch between the inherited productArity = 1 and the synthesized productElementName that throws.

===============================
